src/Service/ProducerSideServiceImpl.cpp

```diff
@@ -356,16 +358,32 @@ void ProducerSideServiceImpl::ReceiveEventsThread(
     uint64_t producer_id, bool* all_events_sent_received) {
   orbit_base::SetCurrentThreadName("PSSI::RcvEvents");
 
-  orbit_grpc_protos::ReceiveCommandsAndSendEventsRequest request;
-  while (stream->Read(&request)) {
+  // Read the ReceiveCommandsAndSendEventsRequest on a protobuf Arena, which prevents memory
+  // allocations for the multiple sub-messages containing ProducerCaptureEvents. Now, when calling
+  // ProducerEventProcessor::ProcessEvent below, passing the ProducerCaptureEvent will result in a
+  // deep copy and the memory allocations we saved before will happen here, plus copies.
+  // Nonetheless, and maybe counterintuitively, the performance improvement we measured is huge. In
+  // particular, ServerReaderWriter::Read seems particularly inefficient without an Arena.
+  google::protobuf::ArenaOptions arena_options;
+  constexpr size_t kArenaInitialBlockSize = 1024 * 1024;
```

Comment:
Is there a good reason for 1M? Do you have an idea how many events are in one message and what's there approximate size?

Reply:
I kept the same value as in `LockFreeBufferCaptureEventProducer::ForwarderThread`. It's basically the same code. From that function, the max number of events is 10000, so I guess I was assuming ~100 B per events (otherwise the default block sizes are quite small).

Reply:
I think you're right, done.